src/test/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/helper/builders/CourseActivity.kt

```diff
@@ -0,0 +1,220 @@
+@Component
+class CourseActivityBuilderRepository(
+  private val courseActivityRepository: CourseActivityRepository? = null,
+  private val agencyLocationRepository: AgencyLocationRepository? = null,
+  private val iepLevelRepository: ReferenceCodeRepository<IEPLevel>? = null,
+  private val agencyInternalLocationRepository: AgencyInternalLocationRepository? = null,
+) {
+  fun save(courseActivity: CourseActivity) = courseActivityRepository?.save(courseActivity)
+  fun lookupAgency(id: String): AgencyLocation = agencyLocationRepository?.findByIdOrNull(id)!!
+  fun lookupIepLevel(code: String): IEPLevel =
+    iepLevelRepository?.findByIdOrNull(ReferenceCode.Pk(IEPLevel.IEP_LEVEL, code))!!
+  fun lookupAgencyInternalLocation(locationId: Long): AgencyInternalLocation? =
+    agencyInternalLocationRepository?.findByIdOrNull(locationId)
+}
+
+@Component
+class CourseActivityBuilderFactory(
+  private val repository: CourseActivityBuilderRepository? = null,
+  private val courseActivityPayRateBuilderFactory: CourseActivityPayRateBuilderFactory = CourseActivityPayRateBuilderFactory(),
+  private val courseScheduleBuilderFactory: CourseScheduleBuilderFactory = CourseScheduleBuilderFactory(),
+  private val courseScheduleRuleBuilderFactory: CourseScheduleRuleBuilderFactory = CourseScheduleRuleBuilderFactory(),
+) {
+  fun builder(): CourseActivityBuilder {
+    return CourseActivityBuilder(
+      repository,
+      courseActivityPayRateBuilderFactory,
+      courseScheduleBuilderFactory,
+      courseScheduleRuleBuilderFactory,
+    )
+  }
+}
+
+class CourseActivityBuilder(
+  val repository: CourseActivityBuilderRepository? = null,
+  val courseActivityPayRateBuilderFactory: CourseActivityPayRateBuilderFactory,
+  val courseScheduleBuilderFactory: CourseScheduleBuilderFactory,
+  val courseScheduleRuleBuilderFactory: CourseScheduleRuleBuilderFactory,
+) : CourseActivityDsl {
+
+  private lateinit var courseActivity: CourseActivity
+
+  fun build(
+    programService: ProgramService,
+    courseActivityId: Long,
+    code: String,
+    prisonId: String,
+    description: String,
+    capacity: Int,
+    active: Boolean,
+    startDate: String,
+    endDate: String?,
+    minimumIncentiveLevelCode: String,
+    internalLocationId: Long?,
+    excludeBankHolidays: Boolean,
+  ): CourseActivity =
+    CourseActivity(
+      courseActivityId = courseActivityId,
+      code = code,
+      program = programService,
+      caseloadId = prisonId,
+      prison = lookupAgency(prisonId),
+      description = description,
+      capacity = capacity,
+      active = active,
+      scheduleStartDate = LocalDate.parse(startDate),
+      scheduleEndDate = endDate?.let { LocalDate.parse(it) },
+      iepLevel = lookupIepLevel(minimumIncentiveLevelCode),
+      internalLocation = internalLocationId?.let { lookupAgencyInternalLocation(it, prisonId) },
+      excludeBankHolidays = excludeBankHolidays,
+    )
+      .let { save(it) }
+      .also { courseActivity = it }
```

Comment:
When we build the course activity we:
* take the ProgramService as passed in from its parent builder
* save the CourseActivity and return it so it can be assigned to a variable in tests
* save a local reference to the CourseActivity so we can pass that to any children